File: source/clibCore/Models/Context/propertyModel.vb

Option Explicit On
Option Strict On

Imports Contensive.Core.Controllers
Imports Contensive.Core.Controllers.genericController

Namespace Contensive.Core.Models.Context
    Public Class propertyModelClass
        '
        Private cpCore As coreClass
        '
        ' visit property cache
        '
        Private propertyCache(,) As String
        Private propertyCache_nameIndex As keyPtrController
        Private propertyCacheLoaded As Boolean = False
        Private propertyCacheCnt As Integer
        Private propertyTypeId As Integer
        '
        '==============================================================================================
        ''' <summary>
        ''' constructor
        ''' </summary>
        ''' <param name="cp"></param>
        ''' <param name="propertyTypeId"></param>
        ''' <remarks></remarks>
        Public Sub New(cpCore As coreClass, propertyTypeId As Integer)
            Me.cpCore = cpCore
            Me.propertyTypeId = propertyTypeId
        End Sub
        '
        '====================================================================================================
        ''' <summary>
        ''' set property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="PropertyValue"></param>
        Public Sub setProperty(ByVal propertyName As String, ByVal PropertyValue As Double)
            setProperty(propertyName, PropertyValue.ToString())
        End Sub
        '
        '====================================================================================================
        ''' <summary>
        ''' set property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="PropertyValue"></param>
        Public Sub setProperty(ByVal propertyName As String, ByVal PropertyValue As Boolean)
            setProperty(propertyName, PropertyValue.ToString())
        End Sub
        '
        '====================================================================================================
        ''' <summary>
        ''' set property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="PropertyValue"></param>
        Public Sub setProperty(ByVal propertyName As String, ByVal PropertyValue As Date)
            setProperty(propertyName, PropertyValue.ToString())
        End Sub
        '
        '====================================================================================================
        ''' <summary>
        ''' set property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="PropertyValue"></param>
        Public Sub setProperty(ByVal propertyName As String, ByVal PropertyValue As Integer)
            setProperty(propertyName, PropertyValue.ToString())
        End Sub
        '
        '====================================================================================================
        ''' <summary>
        ''' set property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="PropertyValue"></param>
        Public Sub setProperty(ByVal propertyName As String, ByVal PropertyValue As String)
            Select Case propertyTypeId
                Case PropertyTypeVisit
                    setProperty(propertyName, PropertyValue, cpCore.authContext.visit.id)
                Case PropertyTypeVisitor
                    setProperty(propertyName, PropertyValue, cpCore.authContext.visitor.ID)
                Case PropertyTypeMember
                    setProperty(propertyName, PropertyValue, cpCore.authContext.user.id)
            End Select
        End Sub
        '
        '====================================================================================================
        ''' <summary>
        ''' set property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="PropertyValue"></param>
        ''' <param name="keyId">keyId is like vistiId, vistorId, userId</param>
        Public Sub setProperty(ByVal propertyName As String, ByVal PropertyValue As String, ByVal keyId As Integer)
            On Error GoTo ErrorTrap 'Const Tn = "SetCSv_VisitProperty" : ''Dim th as integer : th = profileLogMethodEnter(Tn)
            '
            Dim Ptr As Integer
            Dim RecordID As Integer
            Dim CS As Integer
            Dim SQLNow As String
            Dim db As dbController = cpCore.db
            '
            Ptr = -1
            If Not propertyCacheLoaded Then
                Call loadFromDb(keyId)
            End If
            '
            If propertyCacheCnt > 0 Then
                Ptr = propertyCache_nameIndex.getPtr(propertyName)
            End If
            If Ptr < 0 Then
                Ptr = propertyCacheCnt
                propertyCacheCnt = propertyCacheCnt + 1
                ReDim Preserve propertyCache(2, Ptr)
                propertyCache(0, Ptr) = propertyName
                propertyCache(1, Ptr) = PropertyValue
                Call propertyCache_nameIndex.setPtr(propertyName, Ptr)
                '
                ' insert a new property record, get the ID back and save it in cache
                '
                CS = db.cs_insertRecord("Properties", SystemMemberID)
                If db.cs_ok(CS) Then
                    propertyCache(2, Ptr) = db.cs_getText(CS, "ID")
                    Call db.cs_set(CS, "name", propertyName)
                    Call db.cs_set(CS, "FieldValue", PropertyValue)
                    Call db.cs_set(CS, "TypeID", propertyTypeId)
                    Call db.cs_set(CS, "KeyID", CStr(keyId))
                End If
                Call db.cs_Close(CS)
            ElseIf propertyCache(1, Ptr) <> PropertyValue Then
                propertyCache(1, Ptr) = PropertyValue
                RecordID = genericController.EncodeInteger(propertyCache(2, Ptr))
                SQLNow = db.encodeSQLDate(Now)
                '
                ' save the value in the property that was found
                '
                Call db.executeQuery("update ccProperties set FieldValue=" & db.encodeSQLText(PropertyValue) & ",ModifiedDate=" & SQLNow & " where id=" & RecordID)
            End If
            '

            '
            Exit Sub
            '
            ' ----- Error Trap
            '
ErrorTrap:
            cpCore.handleException(New Exception("Unexpected exception"))
        End Sub
        '
        '====================================================================================================
        ''' <summary>
        ''' get a property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getDate(ByVal propertyName As String) As Date
            Return getDate(propertyName, Date.MinValue)
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get a property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getDate(ByVal propertyName As String, ByVal defaultValue As Date) As Date
            Select Case propertyTypeId
                Case PropertyTypeVisit
                    Return getDate(propertyName, defaultValue, cpCore.authContext.visit.id)
                Case PropertyTypeVisitor
                    Return getDate(propertyName, defaultValue, cpCore.authContext.visitor.ID)
                Case PropertyTypeMember
                    Return getDate(propertyName, defaultValue, cpCore.authContext.user.id)
            End Select
            Return Date.MinValue
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get a property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getDate(ByVal propertyName As String, ByVal defaultValue As Date, ByVal keyId As Integer) As Date
            Return genericController.EncodeDate(getText(propertyName, genericController.encodeText(defaultValue), keyId))
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get a property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getNumber(ByVal propertyName As String) As Double
            Return getNumber(propertyName, 0)
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get a property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getNumber(ByVal propertyName As String, ByVal defaultValue As Double) As Double
            Select Case propertyTypeId
                Case PropertyTypeVisit
                    Return getNumber(propertyName, defaultValue, cpCore.authContext.visit.id)
                Case PropertyTypeVisitor
                    Return getNumber(propertyName, defaultValue, cpCore.authContext.visitor.ID)
                Case PropertyTypeMember
                    Return getNumber(propertyName, defaultValue, cpCore.authContext.user.id)
            End Select
            Return 0
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get a boolean property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getNumber(ByVal propertyName As String, ByVal defaultValue As Double, ByVal keyId As Integer) As Double
            Return EncodeNumber(getText(propertyName, genericController.encodeText(defaultValue), keyId))
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get a property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getBoolean(ByVal propertyName As String) As Boolean
            Return getBoolean(propertyName, False)
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get a property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getBoolean(ByVal propertyName As String, ByVal defaultValue As Boolean) As Boolean
            Select Case propertyTypeId
                Case PropertyTypeVisit
                    Return getBoolean(propertyName, defaultValue, cpCore.authContext.visit.id)
                Case PropertyTypeVisitor
                    Return getBoolean(propertyName, defaultValue, cpCore.authContext.visitor.ID)
                Case PropertyTypeMember
                    Return getBoolean(propertyName, defaultValue, cpCore.authContext.user.id)
            End Select
            Return False
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get a boolean property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getBoolean(ByVal propertyName As String, ByVal defaultValue As Boolean, ByVal keyId As Integer) As Boolean
            Return genericController.EncodeBoolean(getText(propertyName, genericController.encodeText(defaultValue), keyId))
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get an integer property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getInteger(ByVal propertyName As String) As Integer
            Return getInteger(propertyName, 0)
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get an integer property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getInteger(ByVal propertyName As String, ByVal defaultValue As Integer) As Integer
            Select Case propertyTypeId
                Case PropertyTypeVisit
                    Return getInteger(propertyName, defaultValue, cpCore.authContext.visit.id)
                Case PropertyTypeVisitor
                    Return getInteger(propertyName, defaultValue, cpCore.authContext.visitor.ID)
                Case PropertyTypeMember
                    Return getInteger(propertyName, defaultValue, cpCore.authContext.user.id)
            End Select
            Return 0
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get an integer property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getInteger(ByVal propertyName As String, ByVal defaultValue As Integer, ByVal keyId As Integer) As Integer
            Return genericController.EncodeInteger(getText(propertyName, genericController.encodeText(defaultValue), keyId))
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get a string property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getText(ByVal propertyName As String) As String
            Return getText(propertyName, "")
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get a string property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getText(ByVal propertyName As String, ByVal defaultValue As String) As String
            Select Case propertyTypeId
                Case PropertyTypeVisit
                    Return getText(propertyName, defaultValue, cpCore.authContext.visit.id)
                Case PropertyTypeVisitor
                    Return getText(propertyName, defaultValue, cpCore.authContext.visitor.ID)
                Case PropertyTypeMember
                    Return getText(propertyName, defaultValue, cpCore.authContext.user.id)
            End Select
            Return ""
        End Function
        '
        '====================================================================================================
        ''' <summary>
        ''' get a string property
        ''' </summary>
        ''' <param name="propertyName"></param>
        ''' <param name="defaultValue"></param>
        ''' <param name="keyId"></param>
        ''' <returns></returns>
        Public Function getText(ByVal propertyName As String, ByVal defaultValue As String, ByVal keyId As Integer) As String
            Dim returnString As String = ""
            Try
                Dim Ptr As Integer
                Dim Found As Boolean
                '
                Ptr = -1
                If Not propertyCacheLoaded Then
                    Call loadFromDb(keyId)
                End If
                '
                If propertyCacheCnt > 0 Then
                    Ptr = propertyCache_nameIndex.getPtr(propertyName)
                    If Ptr >= 0 Then
                        returnString = genericController.encodeText(propertyCache(1, Ptr))
                        Found = True
                    End If
                End If
                '
                If Not Found Then
                    returnString = defaultValue
                    Call setProperty(propertyName, defaultValue, keyId)
                End If
            Catch ex As Exception
                cpCore.handleException(ex) : Throw
            End Try
            Return returnString
        End Function
        '
        '
        '
        Private Sub loadFromDb(ByVal keyId As Integer)
            On Error GoTo ErrorTrap 'Const Tn = "LoadVisitProperties" : ''Dim th as integer : th = profileLogMethodEnter(Tn)
            '
            Dim Name As String
            Dim db As dbController = cpCore.db
            '
            propertyCache_nameIndex = New keyPtrController
            propertyCacheCnt = 0
            '
            Using dt As DataTable = db.executeQuery("select Name,FieldValue,ID from ccProperties where (active<>0)and(TypeID=" & propertyTypeId & ")and(KeyID=" & keyId & ")")
                If dt.Rows.Count > 0 Then
                    propertyCacheCnt = 0
                    ReDim propertyCache(2, dt.Rows.Count - 1)
                    For Each dr As DataRow In dt.Rows
                        Name = genericController.encodeText(dr(0))
                        propertyCache(0, propertyCacheCnt) = Name
                        propertyCache(1, propertyCacheCnt) = genericController.encodeText(dr(1))
                        propertyCache(2, propertyCacheCnt) = genericController.EncodeInteger(dr(2)).ToString
                        Call propertyCache_nameIndex.setPtr(LCase(Name), propertyCacheCnt)
                        propertyCacheCnt += 1
                    Next
                    propertyCacheCnt = dt.Rows.Count
                End If
            End Using
            '
            propertyCacheLoaded = True
            '
            Exit Sub
            '
            ' ----- Error Trap
            '
ErrorTrap:
            cpCore.handleException(New Exception("Unexpected exception"))
        End Sub
    End Class
End Namespace

File: source/clibCoreCs/Models/Context/propertyModel.cs

using System;
using System.Reflection;
using System.Xml;
using System.Diagnostics;
using System.Linq;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using Contensive.Core;
using Contensive.Core.Models.Entity;
using Contensive.Core.Controllers;
using static Contensive.Core.Controllers.genericController;
using static Contensive.Core.constants;
//
namespace Contensive.Core.Models.Context {
    public class propertyModelClass {
        //
        private coreClass cpCore;
        //
        // visit property cache
        //
        private string[,] propertyCache;
        private keyPtrController propertyCache_nameIndex;
        private bool propertyCacheLoaded = false;
        private int propertyCacheCnt;
        private int propertyTypeId;
        //
        //==============================================================================================
        /// <summary>
        /// constructor
        /// </summary>
        /// <param name="cp"></param>
        /// <param name="propertyTypeId"></param>
        /// <remarks></remarks>
        public propertyModelClass(coreClass cpCore, int propertyTypeId) {
            this.cpCore = cpCore;
            this.propertyTypeId = propertyTypeId;
        }
        //
        //====================================================================================================
        /// <summary>
        /// set property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="PropertyValue"></param>
        public void setProperty(string propertyName, double PropertyValue) {
            setProperty(propertyName, PropertyValue.ToString());
        }
        //
        //====================================================================================================
        /// <summary>
        /// set property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="PropertyValue"></param>
        public void setProperty(string propertyName, bool PropertyValue) {
            setProperty(propertyName, PropertyValue.ToString());
        }
        //
        //====================================================================================================
        /// <summary>
        /// set property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="PropertyValue"></param>
        public void setProperty(string propertyName, DateTime PropertyValue) {
            setProperty(propertyName, PropertyValue.ToString());
        }
        //
        //====================================================================================================
        /// <summary>
        /// set property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="PropertyValue"></param>
        public void setProperty(string propertyName, int PropertyValue) {
            setProperty(propertyName, PropertyValue.ToString());
        }
        //
        //====================================================================================================
        /// <summary>
        /// set property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="PropertyValue"></param>
        public void setProperty(string propertyName, string PropertyValue) {
            switch (propertyTypeId) {
                case PropertyTypeVisit:
                    setProperty(propertyName, PropertyValue, cpCore.doc.authContext.visit.id);
                    break;
                case PropertyTypeVisitor:
                    setProperty(propertyName, PropertyValue, cpCore.doc.authContext.visitor.ID);
                    break;
                case PropertyTypeMember:
                    setProperty(propertyName, PropertyValue, cpCore.doc.authContext.user.id);
                    break;
            }
        }
        //
        //====================================================================================================
        /// <summary>
        /// set property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="PropertyValue"></param>
        /// <param name="keyId">keyId is like vistiId, vistorId, userId</param>
        public void setProperty(string propertyName, string PropertyValue, int keyId) {
            try {
                //
                int Ptr = 0;
                int RecordID = 0;
                int CS = 0;
                string SQLNow = null;
                dbController db = cpCore.db;
                //
                Ptr = -1;
                if (!propertyCacheLoaded) {
                    loadFromDb(keyId);
                }
                //
                if (propertyCacheCnt > 0) {
                    Ptr = propertyCache_nameIndex.getPtr(propertyName);
                }
                if (Ptr < 0) {
                    Ptr = propertyCacheCnt;
                    propertyCacheCnt = propertyCacheCnt + 1;
                    //INSTANT C# NOTE: The following block reproduces what 'ReDim Preserve' does behind the scenes in VB:
                    //ORIGINAL LINE: ReDim Preserve propertyCache(2, Ptr)
                    string[,] tempVar = new string[3, Ptr + 1];
                    if (propertyCache != null) {
                        for (int Dimension0 = 0; Dimension0 < propertyCache.GetLength(0); Dimension0++) {
                            int CopyLength = Math.Min(propertyCache.GetLength(1), tempVar.GetLength(1));
                            for (int Dimension1 = 0; Dimension1 < CopyLength; Dimension1++) {
                                tempVar[Dimension0, Dimension1] = propertyCache[Dimension0, Dimension1];
                            }
                        }
                    }
                    propertyCache = tempVar;
                    propertyCache[0, Ptr] = propertyName;
                    propertyCache[1, Ptr] = PropertyValue;
                    propertyCache_nameIndex.setPtr(propertyName, Ptr);
                    //
                    // insert a new property record, get the ID back and save it in cache
                    //
                    CS = db.csInsertRecord("Properties", SystemMemberID);
                    if (db.csOk(CS)) {
                        propertyCache[2, Ptr] = db.csGetText(CS, "ID");
                        db.csSet(CS, "name", propertyName);
                        db.csSet(CS, "FieldValue", PropertyValue);
                        db.csSet(CS, "TypeID", propertyTypeId);
                        db.csSet(CS, "KeyID", keyId.ToString());
                    }
                    db.csClose(ref CS);
                } else if (propertyCache[1, Ptr] != PropertyValue) {
                    propertyCache[1, Ptr] = PropertyValue;
                    RecordID = genericController.EncodeInteger(propertyCache[2, Ptr]);
                    SQLNow = db.encodeSQLDate(DateTime.Now);
                    //
                    // save the value in the property that was found
                    //
                    db.executeQuery("update ccProperties set FieldValue=" + db.encodeSQLText(PropertyValue) + ",ModifiedDate=" + SQLNow + " where id=" + RecordID);
                }
                //

                //
                return;
                //
                // ----- Error Trap
                //
            } catch( Exception ex ) {
                cpCore.handleException(ex);
            }
            //ErrorTrap:
            cpCore.handleException(new Exception("Unexpected exception"));
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public DateTime getDate(string propertyName) {
            return getDate(propertyName, DateTime.MinValue);
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public DateTime getDate(string propertyName, DateTime defaultValue) {
            switch (propertyTypeId) {
                case PropertyTypeVisit:
                    return getDate(propertyName, defaultValue, cpCore.doc.authContext.visit.id);
                case PropertyTypeVisitor:
                    return getDate(propertyName, defaultValue, cpCore.doc.authContext.visitor.ID);
                case PropertyTypeMember:
                    return getDate(propertyName, defaultValue, cpCore.doc.authContext.user.id);
            }
            return DateTime.MinValue;
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public DateTime getDate(string propertyName, DateTime defaultValue, int keyId) {
            return genericController.EncodeDate(getText(propertyName, genericController.encodeText(defaultValue), keyId));
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public double getNumber(string propertyName) {
            return getNumber(propertyName, 0);
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public double getNumber(string propertyName, double defaultValue) {
            switch (propertyTypeId) {
                case PropertyTypeVisit:
                    return getNumber(propertyName, defaultValue, cpCore.doc.authContext.visit.id);
                case PropertyTypeVisitor:
                    return getNumber(propertyName, defaultValue, cpCore.doc.authContext.visitor.ID);
                case PropertyTypeMember:
                    return getNumber(propertyName, defaultValue, cpCore.doc.authContext.user.id);
            }
            return 0;
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a boolean property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public double getNumber(string propertyName, double defaultValue, int keyId) {
            return EncodeNumber(getText(propertyName, genericController.encodeText(defaultValue), keyId));
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public bool getBoolean(string propertyName) {
            return getBoolean(propertyName, false);
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public bool getBoolean(string propertyName, bool defaultValue) {
            switch (propertyTypeId) {
                case PropertyTypeVisit:
                    return getBoolean(propertyName, defaultValue, cpCore.doc.authContext.visit.id);
                case PropertyTypeVisitor:
                    return getBoolean(propertyName, defaultValue, cpCore.doc.authContext.visitor.ID);
                case PropertyTypeMember:
                    return getBoolean(propertyName, defaultValue, cpCore.doc.authContext.user.id);
            }
            return false;
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a boolean property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public bool getBoolean(string propertyName, bool defaultValue, int keyId) {
            return genericController.EncodeBoolean(getText(propertyName, genericController.encodeText(defaultValue), keyId));
        }
        //
        //====================================================================================================
        /// <summary>
        /// get an integer property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public int getInteger(string propertyName) {
            return getInteger(propertyName, 0);
        }
        //
        //====================================================================================================
        /// <summary>
        /// get an integer property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public int getInteger(string propertyName, int defaultValue) {
            switch (propertyTypeId) {
                case PropertyTypeVisit:
                    return getInteger(propertyName, defaultValue, cpCore.doc.authContext.visit.id);
                case PropertyTypeVisitor:
                    return getInteger(propertyName, defaultValue, cpCore.doc.authContext.visitor.ID);
                case PropertyTypeMember:
                    return getInteger(propertyName, defaultValue, cpCore.doc.authContext.user.id);
            }
            return 0;
        }
        //
        //====================================================================================================
        /// <summary>
        /// get an integer property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public int getInteger(string propertyName, int defaultValue, int keyId) {
            return genericController.EncodeInteger(getText(propertyName, genericController.encodeText(defaultValue), keyId));
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a string property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public string getText(string propertyName) {
            return getText(propertyName, "");
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a string property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public string getText(string propertyName, string defaultValue) {
            switch (propertyTypeId) {
                case PropertyTypeVisit:
                    return getText(propertyName, defaultValue, cpCore.doc.authContext.visit.id);
                case PropertyTypeVisitor:
                    return getText(propertyName, defaultValue, cpCore.doc.authContext.visitor.ID);
                case PropertyTypeMember:
                    return getText(propertyName, defaultValue, cpCore.doc.authContext.user.id);
            }
            return "";
        }
        //
        //====================================================================================================
        /// <summary>
        /// get a string property
        /// </summary>
        /// <param name="propertyName"></param>
        /// <param name="defaultValue"></param>
        /// <param name="keyId"></param>
        /// <returns></returns>
        public string getText(string propertyName, string defaultValue, int keyId) {
            string returnString = "";
            try {
                int Ptr = 0;
                bool Found = false;
                //
                Ptr = -1;
                if (!propertyCacheLoaded) {
                    loadFromDb(keyId);
                }
                //
                if (propertyCacheCnt > 0) {
                    Ptr = propertyCache_nameIndex.getPtr(propertyName);
                    if (Ptr >= 0) {
                        returnString = genericController.encodeText(propertyCache[1, Ptr]);
                        Found = true;
                    }
                }
                //
                if (!Found) {
                    returnString = defaultValue;
                    setProperty(propertyName, defaultValue, keyId);
                }
            } catch (Exception ex) {
                cpCore.handleException(ex);
                throw;
            }
            return returnString;
        }
        //
        //
        //
        private void loadFromDb(int keyId) {
            try {
                //
                string Name = null;
                dbController db = cpCore.db;
                //
                propertyCache_nameIndex = new keyPtrController();
                propertyCacheCnt = 0;
                //
                using (DataTable dt = db.executeQuery("select Name,FieldValue,ID from ccProperties where (active<>0)and(TypeID=" + propertyTypeId + ")and(KeyID=" + keyId + ")")) {
                    if (dt.Rows.Count > 0) {
                        propertyCacheCnt = 0;
                        propertyCache = new string[3, dt.Rows.Count];
                        foreach (DataRow dr in dt.Rows) {
                            Name = genericController.encodeText(dr[0]);
                            propertyCache[0, propertyCacheCnt] = Name;
                            propertyCache[1, propertyCacheCnt] = genericController.encodeText(dr[1]);
                            propertyCache[2, propertyCacheCnt] = genericController.EncodeInteger(dr[2]).ToString();
                            propertyCache_nameIndex.setPtr(Name.ToLower(), propertyCacheCnt);
                            propertyCacheCnt += 1;
                        }
                        propertyCacheCnt = dt.Rows.Count;
                    }
                }
                //
                propertyCacheLoaded = true;
                //
                return;
                //
                // ----- Error Trap
                //
            } catch( Exception ex ) {
                cpCore.handleException(ex);
            }
            //ErrorTrap:
            cpCore.handleException(new Exception("Unexpected exception"));
        }
    }
}
